Source/Core/Core/AchievementManager.cpp

```diff
@@ -239,6 +249,38 @@ void AchievementManager::DoFrame()
   }
 }
 
+bool AchievementManager::CanPause()
+{
+  u32 frames_to_next_pause = 0;
+  bool can_pause = rc_client_can_pause(m_client, &frames_to_next_pause);
+  if (!can_pause)
+  {
+    OSD::AddMessage("Cannot spam pausing in hardcore mode.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    OSD::AddMessage(
+        fmt::format("Can pause in {} seconds.", static_cast<float>(frames_to_next_pause) / 60),
+        OSD::Duration::VERY_LONG, OSD::Color::RED);
+  }
+  return can_pause;
+}
+
+void AchievementManager::DoIdle()
+{
+  m_idle_thread = std::make_unique<std::thread>([]() {
+    while (Core::GetState(*AchievementManager::GetInstance().m_system) == Core::State::Paused &&
+           AchievementManager::GetInstance().m_background_execution_allowed)
+    {
+      Common::SleepCurrentThread(1000);
+      std::lock_guard lg{AchievementManager::GetInstance().m_client_lock};
+      auto* client = AchievementManager::GetInstance().m_client;
+      if (!client || !AchievementManager::GetInstance().IsGameLoaded())
+        return;
+      rc_client_idle(client);
```

Comment:
`m_client_lock` is guarding against the other thread reassigning `m_client`, but not against the other thread using the struct pointed to by `m_client`. So I need to ask: Is calling `rc_client_idle` while the other thread is also using `m_client` thread-safe? The rcheevos documentation for `rc_client_idle` doesn't say anything about it as far as I can tell. There is a mutex being locked inside `rc_client_idle`, but I don't know if that means the whole function is safe or not.

Reply:
I've gone over rc_client_idle and my end takeaway is that it is _not_ threadsafe enough for this. As a result, I would prefer to lock every use of m_client, and can use the original m_lock to do so because it's already everywhere.